### PR TITLE
[GSoC] Add preference to allow/warn if using metered connections

### LINKED ISSUE #10706: [Feature request] Option to disable auto sync with metered connections
**Is your feature request related to a problem? Please describe.**
I have auto sync enabled, but I don't want to spend my mobile data with automatic syncs. Mobile data prices can be quite a concern in some countries

**Describe the solution you'd like**
An option to only do auto sync on not metered connections
Extra: a dialog asking if user want to proceed on manual syncs on metered connections

**Describe alternatives you've considered**
None

**Additional context**
JsApi has something similar already #6784

### PR DESCRIPTION
**Pull request by BrayanDSO**
## Purpose / Description
Internet may be expensive 💵💰💲💸

## Fixes
Fixes #10706

## Approach
- Clean things a bit
- Create a "Advanced" category under "Sync", so things get a little more organized with the increasing number of preferences
- Add a preference to allow sync on metered connections (off by default)
     - If not allowed, auto sync under metered connections won't occur + the user will be warned if they trigger a manual sync under a metered connection
     - I'm quite open to string changes

## How Has This Been Tested?

https://user-images.githubusercontent.com/69634269/185769043-30b76c32-5e89-4b27-93b2-156638e26204.mp4

## Checklist
_Please, go through these checks before submitting the PR._

- [X] You have a descriptive commit message with a short title (first line, max 50 chars).
- [X] You have commented your code, particularly in hard-to-understand areas
- [X] You have performed a self-review of your own code
- [X] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Review by lukstbit — DISMISSED** [review]
Looks good with two minor issues.

File: AnkiDroid/src/main/res/values/03-dialogs.xml [hunk 1]
```diff
@@ -257,6 +257,7 @@
         <item quantity="one">An automatic sync may be triggered in %d second</item>
         <item quantity="other">An automatic sync may be triggered in %d seconds</item>
     </plurals>
+    <string name="metered_sync_warning">Your connection is metered. Data transfer may cost money</string>
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/res/values/03-dialogs.xml:260]: Maybe adding a checkbox "Don't warn next time" on the dialog would be good?

[the remark this answers is not in the document]
Comment (lukstbit) [reply AnkiDroid/src/main/res/values/03-dialogs.xml:260]: I would actually prefer the string change because it mentions the setting.
One issue using the checkbox  would be if the user checks "don't warn me" and then selects cancel. At this point the sync would silently fail and could confuse the user(also the material dialogs library allows for a checkbox).

I'm not fixed on the string change so I'm leaving this dialog string issue to the second reviewer.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 2]
```diff
@@ -1563,23 +1565,28 @@ open class DeckPicker :
             Timber.w("User not logged in")
             mPullToSyncWrapper.isRefreshing = false
             showSyncErrorDialog(SyncErrorDialog.DIALOG_USER_NOT_LOGGED_IN_SYNC)
-        } else {
+            return
+        }
+        fun doSync() {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1570]: I'd appreciate a little javadoc on it.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 3]
```diff
@@ -1563,23 +1565,28 @@ open class DeckPicker :
...
-                    Connection.Payload(
-                        arrayOf(
-                            hkey,
-                            preferences.getBoolean("syncFetchesMedia", true),
-                            conflict,
-                            HostNumFactory.getInstance(baseContext)
-                        )
-                    )
-                )
+                val shouldFetchMedia = preferences.getBoolean("syncFetchesMedia", true)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1574]: I'd appreciate for a next time that you do this in multiple commit. 
* Extracting variables in a commit, 
* extracting doSync
* then taking into account whether it's metered

We can put it in a separate PR and squash it, but it would really help to review and save me time

File: AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt [hunk 4]
```diff
@@ -398,14 +398,7 @@ open class AnkiDroidJsAPI(private val activity: AbstractFlashcardViewer) {
...
     fun ankiIsActiveNetworkMetered(): Boolean {
-        return try {
-            val cm = AnkiDroidApp.getInstance().applicationContext
-                .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
-            cm.isActiveNetworkMetered
-        } catch (e: Exception) {
-            Timber.w(e, "Exception obtaining metered connection - assuming metered connection")
-            true
-        }
+        return NetworkUtils.isActiveNetworkMetered()
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/AnkiDroidJsAPI.kt:401]: I particularly prefer functions stating `return`, so I'll keep that

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+import timber.log.Timber
+
+object NetworkUtils {
+    fun isActiveNetworkMetered(): Boolean {
+        return try {
+            val cm = AnkiDroidApp.getInstance().applicationContext
+                .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+            cm.isActiveNetworkMetered
+        } catch (e: Exception) {
+            Timber.w(e, "Exception obtaining metered connection - assuming metered connection")
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt]: I actually just copied and pasted the method, a 100% straight extraction. It wasn't my goal to refactor it on this PR.

But I agree if your points. Looking on the source code, I only expect it to return an error in the case there isn't a specific permission on the manifest. AnkiDroid's manifest has it, so I don't think that the try catch is necessary.

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt (hunk 3 above)
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1576]: `doSync` is mostly only useful inside `sync()`, so to reduce boilerplate, I've kept it as a nested function. Added a javadoc to make it more clear. Let me know if that is fine

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            true
+        }
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt]: After I changed isActiveNetworkMetered, creating a enum may not be necessary anymore, as I've removed the try-catch

**Comment by mikehardy** [comment]
Took a conflict just now, and I'm not sure what the state on it is? Is it ready for re-review from previous reviewers?
Only asking because I'm scouring for things to merge before doing a new build and this caught my eye :eyes:

**Comment by BrayanDSO** [comment]
Yes, it is ready for review. Just rebased to solve conflicts. Only the last 3 commits are new.

File: AnkiDroid/src/main/res/values/03-dialogs.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (lukstbit) [reply AnkiDroid/src/main/res/values/03-dialogs.xml:260]: @BrayanDSO Have you decided on this? The latest changes look ok.

[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/res/values/03-dialogs.xml:260]: I think I'll keep it with the current way. 

In the future someone else can create a new string and concatenate with this one. Or use a "Don't warn me" checkbox. Or use a button that leads directly to the settings screen and highlight the preference. Or anything else

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 7]
```diff
@@ -1567,25 +1570,30 @@ open class DeckPicker :
...
             showSyncErrorDialog(SyncErrorDialog.DIALOG_USER_NOT_LOGGED_IN_SYNC)
-        } else {
-            val syncMedia = preferences.getBoolean("syncFetchesMedia", true)
-
+            return
+        }
+        /** Nested function that makes the connection to
+         * the sync server and starts syncing the data */
+        fun doSync() {
+            val shouldFetchMedia = preferences.getBoolean("syncFetchesMedia", true)
```
Comment (oakkitten) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1578]: Why rename `syncMedia`?

Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1578]: There wasn't a syncMedia when I first did the PR. Probably got it after the rebase. It doesn't seem relevant, so I'll spare myself the time of changing it back

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1578]: I'm ambivalent on the name as long as it all still works but @oakkitten was just *all* over this code making it work well for custom sync servers, so making sure it works after all the custom sync server work is important. From that perspective it may not be resolved in their mind, and that's legitimate from perspective of "time spent wrestling with this mess"

Comment (oakkitten) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1578]: I am one of those people who spend ⅔ of their coding time naming and renaming things 🙈

Anyway, `syncFetchesMedia` implies that only fetching is happening and not uploading, which is confusing. I suppose at some point this setting did something else. So I went with `syncMedia`, which is also the name of the called function parameter.

Oh well it's not really important.

Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt]: now definitely resolved :laughing: - last commit moves it

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 8]
```diff
@@ -20,15 +20,13 @@ import android.net.ConnectivityManager
...
-            true
-        }
+fun isActiveNetworkMetered(): Boolean {
+    return try {
+        val cm = AnkiDroidApp.getInstance().applicationContext
+            .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+        cm.isActiveNetworkMetered
+    } catch (e: Exception) {
+        Timber.w(e, "Exception obtaining metered connection - assuming metered connection")
+        true
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt]: @Arthur-Milchior in your review you wanted an enum, with this refactor here to a simple util method, I think this is clean and correct and just binary. The assertion it is correct is that from the perspective of consuming people's money (using metered data) defaulting to a cautious stance of assuming it is metered when we catch an (almost entirely impossible) exception is the right one.

So I think this may answer your review comments well

**Review by mikehardy — APPROVED** [review]
This looks good to me, my limited data friends (and myself on occasion - if I'm syncing a big deck) will appreciate this.

File: AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt [hunk 9]
```diff
@@ -18,15 +18,16 @@ package com.ichi2.utils
 import android.content.Context
 import android.net.ConnectivityManager
 import com.ichi2.anki.AnkiDroidApp
-import timber.log.Timber
 
+/**
+ * @return whether the user connection is metered
+ */
 fun isActiveNetworkMetered(): Boolean {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/utils/NetworkUtils.kt:25]: Nit: we may avoid all variables and directly use an experssion

```kotlin
fun isActiveNetworkMetered() = 
    AnkiDroidApp
        .getInstance()
        .applicationContext
        .getSystemService(Context.CONNECTIVITY_SERVICE)
        .let {it as ConnectivityManager}
        .isActiveNetworkMetered
```

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 10]
```diff
@@ -1572,6 +1572,8 @@ open class DeckPicker :
             showSyncErrorDialog(SyncErrorDialog.DIALOG_USER_NOT_LOGGED_IN_SYNC)
             return
         }
+        /** Nested function that makes the connection to
+         * the sync server and starts syncing the data */
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:1576]: I'd consider adding it in the commit you created the method.

**Review by Arthur-Milchior — APPROVED** [review]
Generally approve.
I believe we could save two commits. 
If we squash, feel free to squash. 
If we don't, then feel free to merge as soon as the correction is done

File: AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt [hunk 11]
```diff
@@ -924,14 +925,15 @@ open class DeckPicker :
     private fun automaticSync() {
         val preferences = AnkiDroidApp.getSharedPrefs(baseContext)
 
-        // Check whether the option is selected, the user is signed in and last sync was AUTOMATIC_SYNC_TIME ago
-        // (currently 10 minutes)
+        // Check whether the option is selected, the user is signed in, last sync was AUTOMATIC_SYNC_TIME ago
+        // (currently 10 minutes), and is not under a metered connection (if not allowed by preference)
```
[the remark this answers is not in the document]
Comment (BrayanDSO) [reply AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.kt:928]: Thanks for the heads up. Changed the commit message to:

> Add "Allow sync on metered connections" preference
If disabled, auto-sync isn't triggered on metered connections and the user is asked if they want to continue if they trigger a manual sync
If enabled, sync occurs normally

**Comment by mikehardy** [comment]
I think we are good to go with the standard rebase merge here? I'll do a strings run now
